python/ray/workflow/workflow_state.py

```diff
@@ -0,0 +1,290 @@
+import asyncio
+
+from collections import deque, defaultdict
+import dataclasses
+from dataclasses import field
+import logging
+from typing import List, Dict, Optional, Set, Deque, Callable
+
+import ray
+from ray.workflow.common import (
+    Workflow,
+    StepID,
+    WorkflowRef,
+    WorkflowStepRuntimeOptions,
+)
+from ray.workflow.workflow_context import WorkflowStepContext
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class TaskExecutionMetadata:
+    submit_time: Optional[float] = None
+    finish_time: Optional[float] = None
+    output_size: Optional[int] = None
+
+    @property
+    def duration(self):
+        return self.finish_time - self.submit_time
+
+
+@dataclasses.dataclass
+class WorkflowExecutionState:
+    """The execution state of a workflow. This dataclass helps with observation
+    and debugging."""
+
+    # -------------------------------- dependencies -------------------------------- #
+
+    # The reversed dependency mapping from downstream tasks to upstream tasks.
+    in_dependency: Dict[StepID, List[StepID]] = field(default_factory=dict)
+    # The dependency mapping from upstream tasks to downstream tasks.
+    out_dependency: Dict[StepID, List[StepID]] = field(
+        default_factory=lambda: defaultdict(list)
+    )
+    # The mapping from a task to its immediate continuation.
```

Comment:
Why do we treat continuations differently from normal step dependencies?

Reply:
we can reuse the normal dependencies for continuations, but they are very different: continuations does not provide meaningful outputs until the lineage of continuations completes (the last continuation that returns a value instead of a DAG). This affects a lot of implementations. Using a different data structure makes it easier for implementation. Also DAGs in the continuation shares the normal step dependencies data structure.